docs/developers/style-guide.md

```diff
@@ -2,6 +2,94 @@
 title: Style Guide
```

Comment:
Can you run markdownlint's auto-formatter over this file, please?

Reply:
How do I do that?

Reply:
I installed the markdownlint extension in vscode and am since then able to run auto formatting of md files with it. I googled a bit, and it seem that running markdownlint's auto-formatter outside of vscode it's necessary to install a javascript tool:

```
npm install -g markdownlint-cli2 markdownlint-cli2-config-prettier
```

Then you can run

```
markdownlint-cli2 "docs/**/*.md" --fix
```

to fix everything in a folder. It uses the `.markdownlint.yaml` that we've got in the `docs` folder.

Reply:
Is there a way to make things simpler? I would strongly prefer something handled with pre-commit. Otherwise it's not enforceable.
If we keep this it needs to be documented (this style guide?)

Reply:
you've just got to install this npm package in your environment, which requires nodejs. this can be done in CI/docker as well.
there is no python version of markdownlint.

Reply:
pre-commit hook would look like this:

```
  - repo: https://github.com/DavidAnson/markdownlint-cli2
    rev: v0.16.0
    hooks:
      - id: markdownlint-cli2
        name: markdownlint
        entry: markdownlint-cli2
        args: ["--fix"]  # for actually fixing things
        language: node
        types: [markdown]
```

I previously suggested adding the ruby version of markdownlint, but that one doesn't support auto-fixing

Reply:
This doesn't seem to use the `markdownmlint.yaml` config,and doesn't fix everything (not sure if the errors would go away with the config)

Reply:
weird. I just tried running this tool myself in the terminal, and it worked as intended:

```
$ markdownlint-cli2 docs/developers/style-guide.md 
markdownlint-cli2 v0.14.0 (markdownlint v0.35.0)
Finding: docs/developers/style-guide.md
Linting: 1 file(s)
Summary: 21 error(s)
docs/developers/style-guide.md:28 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/developers/style-guide.md:38:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:40:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:42:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:43:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:52:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:54:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:55:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:56:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:57:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:60:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:66 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/developers/style-guide.md:77:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:79:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:82:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:90:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:92:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:93:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:97:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:98:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
docs/developers/style-guide.md:99:1 MD030/list-marker-space Spaces after list markers [Expected: 3; Actual: 1]
```

found 21 linting errors

```
$ markdownlint-cli2 docs/developers/style-guide.md --fix
markdownlint-cli2 v0.14.0 (markdownlint v0.35.0)
Finding: docs/developers/style-guide.md
Linting: 1 file(s)
Summary: 2 error(s)
docs/developers/style-guide.md:28 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/developers/style-guide.md:66 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
```

so it fixed all except 2 errors in the end.

Reply:
<img width="1228" alt="Screenshot 2024-12-18 at 5 19 46 PM" src="https://github.com/user-attachments/assets/ada1d9da-d4c3-493e-baa9-9aad52c574b8" />

^^^ these errors are because of the special mkdocs syntax that markdownlint doesn't expect. I'm looking for a workaround.

Reply:
you can use `<!-- markdownlint-disable-next-line -->` or the pair of `<!-- markdownlint-disable -->` and `<!-- markdownlint-enable -->` to make exemptions to linting

Reply:
Hmm if it was just about this comment I would be fine with it, but I tried running the command and in addition to making a whole lot of changes (had another look, seems like these are mostly outside the docs directory) it returned a whole lot of errors. I assumed this was because of wrong config, but maybe that's not the case. I'll fix the list indents, but we can't really add the pre-commit or enforce markdownlint until this is fixed.
```
markdownlint-cli2 v0.16.0 (markdownlint v0.36.1)
Finding: docs/recipes/continue-training-llama-8b.md docs/developers/contributing.md docs/about-us.md README.md
Linting: 4 file(s)
Summary: 6 error(s)
README.md:1:1 MD033/no-inline-html Inline HTML [Element: div]
README.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "<div align="center" style="mar..."]
README.md:3:1 MD033/no-inline-html Inline HTML [Element: img]
README.md:9 MD036/no-emphasis-as-heading Emphasis used instead of a heading [Context: "Accelerating your LLM training..."]
README.md:110 MD024/no-duplicate-heading Multiple headings with the same content [Context: "Prerequisites"]
README.md:116 MD024/no-duplicate-heading Multiple headings with the same content [Context: "Steps"]
markdownlint-cli2 v0.16.0 (markdownlint v0.36.1)
Finding: docs/quick-start.md .github/PULL_REQUEST_TEMPLATE.md .github/ISSUE_TEMPLATE/bug_report.md CODE_OF_CONDUCT.md
Linting: 4 file(s)
Summary: 37 error(s)
.github/ISSUE_TEMPLATE/bug_report.md:10 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "🐞 Describe the Bug"]
.github/ISSUE_TEMPLATE/bug_report.md:14 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "🔄 Steps to Reproduce"]
.github/ISSUE_TEMPLATE/bug_report.md:37 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "🎯 Expected Behavior"]
.github/ISSUE_TEMPLATE/bug_report.md:41 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "📜 Environment Information"]
.github/ISSUE_TEMPLATE/bug_report.md:108 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "📝 Additional Context"]
docs/quick-start.md:31 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:46 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:106 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:110 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:184 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:188 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:196 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:205 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:231 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:252 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:276 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:284 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:292 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:333 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:471 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:475 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:481 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:535 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:615 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:619 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:630 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:641 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:682 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:834 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:840 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:846 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:854 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:879 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:902 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:933 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:947 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/quick-start.md:963:1 MD053/link-image-reference-definitions Link and image reference definitions should be needed [Unused link or image reference definition: "^peak-tflops"] [Context: "[^peak-tflops]:"]
markdownlint-cli2 v0.16.0 (markdownlint v0.36.1)
Finding: docs/help.md docs/index.md .github/ISSUE_TEMPLATE/feature_request.md docs/developers/style-guide.md
Linting: 4 file(s)
Summary: 7 error(s)
.github/ISSUE_TEMPLATE/feature_request.md:10 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "🧐 Problem Description"]
.github/ISSUE_TEMPLATE/feature_request.md:15 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "💡 Proposed Solution"]
.github/ISSUE_TEMPLATE/feature_request.md:20 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "🔄 Alternatives Considered"]
.github/ISSUE_TEMPLATE/feature_request.md:25 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "📈 Potential Benefits"]
.github/ISSUE_TEMPLATE/feature_request.md:30 MD025/single-title/single-h1 Multiple top-level headings in the same document [Context: "📝 Additional Context"]
docs/developers/style-guide.md:28 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/developers/style-guide.md:66 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
markdownlint-cli2 v0.16.0 (markdownlint v0.36.1)
Finding: docs/recipes/upcycle-llama-3b-to-moe.md fast_llm/models/custom/readme.md docs/join-us.md CONTRIBUTING.md
Linting: 4 file(s)
Summary: 0 error(s)
markdownlint-cli2 v0.16.0 (markdownlint v0.36.1)
Finding: docs/README.md docs/success-stories/starcoder-2.md docs/reference/configuration.md docs/developers/dev-practices.md
Linting: 4 file(s)
Summary: 0 error(s)
markdownlint-cli2 v0.16.0 (markdownlint v0.36.1)
Finding: SECURITY.md docs/recipes/train-llama-8b.md docs/license.md docs/recipes/data-preparation.md
Linting: 4 file(s)
Summary: 5 error(s)
docs/recipes/data-preparation.md:44 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/recipes/data-preparation.md:99 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/recipes/data-preparation.md:107 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/recipes/data-preparation.md:117 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/recipes/data-preparation.md:156 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
```

Reply:
Is there a way to turn these into a warning and do the commit anyway in case of failure? I guess errors like this need to be addressed for PRs, but I don't want this to block every single commit...

Reply:
not sure, this would be great though!